core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1644,23 +1644,32 @@ object Stream {
     def concurrently[O2](that: Stream[F, O2])(implicit F: Effect[F],
                                               ec: ExecutionContext): Stream[F, O] =
       Stream.eval(async.promise[F, Unit]).flatMap { interruptR =>
-        Stream.eval(async.promise[F, Unit]).flatMap { doneR =>
+        Stream.eval(async.promise[F, Option[Throwable]]).flatMap { doneR =>
           Stream.eval(async.promise[F, Throwable]).flatMap { interruptL =>
             def runR =
               that
                 .interruptWhen(interruptR.get.attempt)
                 .compile
                 .drain
                 .attempt
-                .flatMap {
-                  case Right(_)  => doneR.complete(())
-                  case Left(err) => interruptL.complete(err) *> doneR.complete(())
+                .map { _.left.toOption }
+                .flatMap { r =>
+                  doneR.complete(r) >> // to prevent deadlock, done must be signalled before `interruptL`
+                    r.fold(F.pure(()))(interruptL.complete)
                 }
 
+            // There is slight chance that interruption in case of failure will arrive later than
+            // `self` terminates.
+            // To prevent such interruption to be `swallowed` we append stream, that results in
+            // evaluation of the result.
             Stream.eval(async.fork(runR)) >>
               self
                 .interruptWhen(interruptL.get.map(Left(_): Either[Throwable, Unit]))
```

Comment:
`.get.map(Either.left)` to remove that type annotation?

Reply:
ok thanks